source/slang/slang-lower-to-ir.cpp

```diff
@@ -1827,28 +1882,186 @@ LoweredValInfo createVar(
+            // copy-in/copy-out for a `ref` parameter, so we just
+            // pass in the actual pointer.
+            //
+            IRInst* argPtr = getAddress(context, argVal, loc);
+            addInArg(context, ioArgs, LoweredValInfo::simple(argPtr));
+        }
+        break;
+
+    case kParameterDirection_Out:
+    case kParameterDirection_InOut:
+        {
+            // According to our "calling convention" we need to
+            // pass a pointer into the callee.
+            //
+            // A naive approach would be to just take the address
+            // of `loweredArg` above and pass it in, but that
+            // has two issues:
+            //
+            // 1. The l-value might not be something that has a single
+            //    well-defined "address" (e.g., `foo.xzy`).
+            //
+            // 2. The l-value argument might actually alias some other
+            //    storage that the callee will access (e.g., we are
+            //    passing in a global variable, or two `out` parameters
+            //    are being passed the same location in an array).
+            //
+            // In each of these cases, the safe option is to create
+            // a temporary variable to use for argument-passing,
+            // and then do copy-in/copy-out around the call.
+            //
+            // TODO: We should consider ruling out case (2) as undefined
+            // behavior, and specify that whether `inout` and `out` are
+            // handled via copy-in-copy-out or by-reference parameter
+            // passing is an implementation detail. That would allow
+            // us to avoid introducing a copy except where it is required
+            // for the semantics of (1).
+            //
+            // TODO: We should confirm whether such a change will make
+            // it harder to create SSA values for variables that get
+            // used with `out` or `inout` parameters.
+
+            LoweredValInfo tempVar = createVar(context, paramType);
+
+            // If the parameter is `in out` or `inout`, then we need
+            // to ensure that we pass in the original value stored
+            // in the argument, which we accomplish by assigning
+            // from the l-value to our temp.
+            if(paramDirection == kParameterDirection_InOut)
+            {
+                assign(context, tempVar, argVal);
+            }
+
+            // Now we can pass the address of the temporary variable
+            // to the callee as the actual argument for the `in out`
+            SLANG_ASSERT(tempVar.flavor == LoweredValInfo::Flavor::Ptr);
+            IRInst* tempPtr = getAddress(context, tempVar, loc);
+            addInArg(context, ioArgs, LoweredValInfo::simple(tempPtr));
+
+            // Finally, after the call we will need
+            // to copy in the other direction: from our
+            // temp back to the original l-value.
+            OutArgumentFixup fixup;
```

Comment:
For properties this can lead to potential performance issues if calling a function with an `inout` parameter that does not actually writes to it. Given the current lowering logic, it is difficult to optimize out the call to `set` in future passes even if the function is inlined, because at that point it is too late to know that the `set` call is related with the `out` parameter.

Reply:
I meant to follow up on this. The code here is old code that just got moved around in this change, so it wasn't in scope to change the details as part of this PR.

That said, I'm not sure I follow the details of your comments, so I don't get which option you are arguing for or against here. Right now at call sites to things with `inout` parameters we take input like:

```
foo(someVal.someField);
```

and translate it to something like (in C++ pseduo-code):

```
auto tmp = someVal.someField; copy from storage to temporary
foo(&tmp); // pass temporary by reference
someVal.someField = tmp; // copy from temporary back to storage
```

The comments associated with the code mention that *if* we can declare by fiat that the language doesn't support a function mutating a storage location that aliases with an `inout` argument (which I think we can), then we can translate this to the following:

```
foo(&someVal.someField); // pass storage directly by reference
```

This latter translation has the benefit of eliminating some copies, but it potentially inhibits breaking up lifetimes in our SSA pass because our analysis isn't smart enough to know that `foo()` won't (and cannot) allow the address that was passed into "escape." Keeping the temporary around except when true by-reference passing is required helps us make sure SSA can do its job, but leaves us with extra temporaries.

Ideally we could eventually have a virtual "register allocation" pass in the IR that collapses independent local variables together when their lifetimes don't overlap (if doing so would eliminate moves between them), and similarly decides to allocate SSA temporaries into those same locations for the purposes of code emit when doing so would eliminate copies.

Anyway, that is all further off...

In the specific case of `property` declarations, the same idiom currently translates to something even more silly:

```
auto tmpProp = someProperty_get(someVal);
foo(&tmp);
auto tmpVal = someVal;
someProperty_set(&tmpVal, tmpProp);
someVal = tmpVal;
```

The reason for the extra verbiage is that the setter for `someProperty` is also going to have an `inout` parameter (for `this`) and so the logic for introducing these copies will apply to that too.

That is clearly a bit of a mess, and we will be reliant on downstream optimization passes (or ones we introduce to Slang) to clean up the resulting logic into something more idiomatic.

Even if we eliminate the extra copies due to our handling of `inout`, I think your point was that we still end up with:

```
auto tmpProp = someProperty_get(someVal);
foo(&tmpProp);
somePropety_set(&someVal, tmpProp);
```

That is, we cannot eliminate the temporary variable for `someProp` because the property itself might not have a fixed address that can be passed along to `foo`.

Your point seems to be that we cannot eliminate the call to `someProperty_set` even if we discover that `foo()` never writes to that `inout` parameter, because the compiler wouldn't have a way to know that assigning back `tmpProp` would do nothing (and indeed there is no reason for the compiler to require/assume that a setter can't have observable side effects).

AFAICT there are two basic ways to deal with that specific concern:

1. We can implement true call-by-reference and have by-reference accessors. It turns out we actually have this in Slang, although it is only supported for Slang stdlib code. Operations can have `ref` parameters instead of just `inout`, such that by-reference parameter passing is guaranteed/required, and subscripts/properties can have `ref` accessors that simply return a reference to storage instead of needing paired setters/getters. Exposing these as first-class constructs for users would require some non-trivial work, but could probably be done.

2. We could implement "call-by-name" parameter passing, so that instead of receiving a pointer (as in the current Slang IR) the IR version of a function with an `inout` parameter receives a kind of object; something like an `IReadWriteCell<T>` that it can use to read and write the argument, invoking the getter/setter logic on each such access (so every line of code in `foo()` that writes to its `inout` parameter would invoke the full setter logic).

Implementing call-by-name would be a bit of a mess in the implementation, and would seemingly move our design further away from HLSL (which is all based on copy-in/copy-out). I honestly don't think of call-by-name as a realistic option.

There is a remaining option in this space that addresses some of the overhead of getter/setter pairs, but that we aren't ready to try and tackle. As usual, this is an idea coming from the Swift world, where they have `modify` accessors a not-yet-final feature to go along with `get` and `set`. The idea of a `modify` accessors is that it is a coroutine, and the value it `yield`s is a reference to storage.

A `modify` accessor is thus a lot like the current `ref` accessor feature in Slang (they both expose a reference to storage to the caller), and both would allow the code generation at a call site to directly use the storage location that was returned/yielded to pass to a function with an `inout` parameter. The key difference is that because it is a coroutine, a `modify` accessor can have code that logically runs *after* the operation that needed access to the storage, which can handle cleanup, writing back to another location, validation, etc.

I like to think that `property` declarations as we are building them right now (with only `get` and `set`) are still a clear syntactic improvement over traditional getter/setter function pairs, even if we don't address some of these thornier challenges. We can hopefully make them perform better over time, but it is reasonable to say that in code that cares a lot about fine-grained performance details for `inout` parameter passing, programmers should favor ordinary fields over `property` declarations for now.

Reply:
I did not mean that we should address the issue now. The comment is only there to remind us that this could be a place for potential performance improvement. A possible alternative to `ref` is we can have functions to take an `Assignable<T>` instead of `inout<T>`, where `Assinable<T>` provides `get` and `set` methods, and the compiler can synthesize such an `Assignable` object for properties or ordinary fields.